e2e/pages/vm-sidebar.po.ts

```diff
@@ -171,4 +172,142 @@ export class VMSidebar extends CloudstackUiPage {
+    ).getText();
+  }
+
+  checkWarning() {
+    element(by.css('.message.warning.ng-star-inserted')).isPresent();
+    return element(by.css('.message.warning.ng-star-inserted')).getText();
+  }
+
+  isEnabledChangeSO() {
+    expect(
+      element(by.css('cs-service-offering-details button[ng-reflect-disabled=true]')),
+    ).toBeTruthy('Change SO button is enabled');
+  }
+
+  isEnabledAddAffinityGroup() {
+    expect(element(by.css('cs-affinity-group button[ng-reflect-disabled=true]'))).toBeTruthy(
+      'Affinity group button is enabled',
+    );
+  }
+
+  isEnabledAddSSHKey() {
+    expect(element(by.css('cs-vm-ssh-keypair button[ng-reflect-disabled=true]'))).toBeTruthy(
+      'Add SHH Key is enabled',
+    );
+  }
+
+  clickChangeSSHKey() {
+    const edit = element(by.css('cs-vm-ssh-keypair button'));
+    const EC = protractor.ExpectedConditions;
+    browser.wait(EC.elementToBeClickable(edit), 2000, 'No edit Group button').then(() => {
+      edit.click();
+      this.waitDialogModal();
+    });
+  }
+
+  selectSSHKey(index) {
+    element(by.css('cs-ssh-keypair-reset mat-select')).click();
+    element
+      .all(by.css('mat-option'))
+      .get(index)
+      .click();
+    return element
+      .all(by.css('mat-option span'))
+      .get(index)
+      .getText();
+  }
+
+  getSHHKey() {
+    return element(by.css('cs-vm-ssh-keypair span')).getText();
+  }
+
+  checkResetWarning() {
+    return browser.wait(
+      protractor.ExpectedConditions.presenceOf(element(by.css('.mat-dialog-content.notification'))),
+      5000,
+    );
+  }
+
+  isEnabledCreateAffGroupButton() {
+    return element(by.css('.add-rule-button.mat-icon-button')).isEnabled();
+  }
+
+  setIncorrectAffGroupName(name) {
+    element(by.css('input[formControlName=name]')).sendKeys(name);
+  }
+
+  getErrorText() {
+    element(by.css('cs-affinity-group-selector mat-card-header')).click();
+    const EC = browser.ExpectedConditions;
+    browser.wait(
+      EC.presenceOf(element(by.css('.mat-error.ng-star-inserted'))),
+      5000,
+      'Error message is not present',
+    );
+    return element(by.css('.mat-error.ng-star-inserted')).getText();
+  }
+
+  clickDeleteAffGroup() {
+    const del = element(by.css('cs-affinity-group .row.ng-star-inserted button'));
+    const EC = protractor.ExpectedConditions;
+    browser.wait(EC.elementToBeClickable(del), 2000, 'does not clicable button').then(() => {
+      del.click();
+    });
+  }
+  getNotificationText() {
+    return element(by.css('cs-confirm-dialog .mat-dialog-content.notification')).getText();
+  }
+
+  checkNoAffGroup() {
+    browser.wait(
+      protractor.ExpectedConditions.presenceOf(
+        element(by.css('cs-affinity-group .mat-card-content-container div')),
+      ),
+      5000,
+    );
+    return element(by.css('cs-affinity-group .mat-card-content-container div')).getText();
+  }
+
+  waitSidebar() {
+    const EC = browser.ExpectedConditions;
```

Comment:
You declare an EU constant again each time before use.
Set this constant once for the entire document, if possible.